06x-new-kafka-roller.md

```diff
@@ -0,0 +1,331 @@
+9. It sends a request to describe all the topic partitions, and the following list of topics is returned:
+```
+topic("topic-A"), Replicas(9, 10, 11), ISR(9, 10)
+topic("topic-B"), Replicas(6, 7, 8), ISR(6, 7, 8)
+topic("topic-C"), Replicas(10, 8, 6), ISR(10, 8, 6)
+topic("topic-D"), Replicas(7, 9, 11), ISR(7, 9, 11)
+topic("topic-E"), Replicas(6, 10, 11), ISR(6, 10, 11)
+```
+They are configured with `min.insync.replica` of 2.
+
+10. The roller batches the nodes that do not have any topic partition in common and the following batches are created: 
+- (11, 8) - `broker-11` and `broker-8` do not share any topic partitions.
+- (7) - `broker-7` and `broker-10` do not share any topic partitions, however topic-A is at min ISR, therefore 10 cannot be restarted and is removed from the batch.
+- (6) - `broker-6` and `broker-9` do not share any topic partitions, however topic-A is at min ISR, therefore 9 cannot be restarted and is removed from the batch.
+
+11. The roller picks the largest batch containing `broker-11` and `broker-8` and restarts them. It waits forthe nodes to have `SERVING` and then `LEADING_ALL_PREFERRED` state.
+12. It then restarts the batch containing only `broker-7` and restart it. It waits for it to have `SERVING` and then `LEADING_ALL_PREFERRED` state.
+13. It then restarts the batch containing only `broker-6` and restart it. It times out waiting for it to have `SERVING` state because it's still performing log recovery.
+14. The roller retries waiting for `broker-6` to have `SERVING` state for a number of times and results in the following context:
+```
+         nodeRef: broker-6/6
+         nodeRoles: broker
+         state: RECOVERING
+         lastTransition: 987456
+         reason:
+         numRestartAttempts: 1
+         numReconfigAttempts: 0
+         numRetries: 10
+```
+15. The `max_retries` of 10 is reached for `broker-6`, therefore the roller throws `UnrestartableNodesException` and the reconciliation fails. The operator logs the number of remaining segments and logs to recover.
+16. When the next reconciliation starts, all the nodes are observed and their contexts are updated. `broker-6` node has finished performing log recovery therefore have `SERVING` state. All nodes have no reason to restart. 
+17. Broker nodes are checked if their configurations have been updated. `min.insync.replica` has been updated to 1 therefore the roller sends a request containing the configuration update to the brokers and the transitions nodes' state to `RECONFIGURED`. 
+18. Observe the broker nodes again, and check if they have `LEADING_ALL_PREFERRED` state. 
+19. All nodes have `SERVING` or `LEADING_ALL_PREFERRED` and no reason to restart.
+20. The reconciliation completes successfully.
+
+
+### Switching from the old KafkaRoller to the new KafkaRoller
+
+The new KafkaRoller will only work with KRaft clusters therefore when running in Zookeeper mode, the current KafkaRoller will be used.
```

Comment:
This would be done simply when a user deploys the KRaft cluster we will use `KafkaRoller v2` or some feature gates be needed?  I assume the first option, so maybe we should clearly state here, which path we are gonna take :)

Reply:
No feature gate is introduced. The paragraph below it explains that KafkaReconciler will switch to the new roller based on KafkaMetadataState of Kafka CR. Do you think it's not clear enough?